compiler/src/dmd/clone.d

```diff
@@ -327,10 +327,19 @@ FuncDeclaration buildOpAssign(StructDeclaration sd, Scope* sc)
         auto e2 = new BlitExp(loc, new VarExp(loc, swap), new ThisExp(loc));
         auto e3 = new BlitExp(loc, new ThisExp(loc), new IdentifierExp(loc, Id.p));
 
-        /* Instead of running the destructor on s, run it
-         * on swap. This avoids needing to copy swap back in to s.
-         */
-        auto e4 = new CallExp(loc, new DotVarExp(loc, new VarExp(loc, swap), sd.dtor, false));
+        Expression e4;
+        if (target.isCalleeDestroyingArgs(tf))
+        {   /* callee destroys s
+             * Instead of running the destructor on s, run it
+             * on swap.
+             */
+            e4 = new CallExp(loc, new DotVarExp(loc, new VarExp(loc, swap), sd.dtor, false));
+        }
+        else
+        {   /* caller destroys s, so copy contents of swap back into s
+             */
+            e4 = new BlitExp(loc, new IdentifierExp(loc, Id.p), new VarExp(loc, swap));
+        }
```

Comment:
It would be good if (in this PR is fine)  there was a visual description (pseudocode) of what this is generating before and after this change, as its hard to put this into context.

Going off the comment alone

> caller destroys s, so copy contents of swap back into s

Makes me want to question why are we making a temp copy in the first place.

Reply:
The robust swap algorithm to swap and b is:

```
tmp = a;
a = b;
b = tmp;
```

Reply:
As to pseudocode, it is there in the comments preceding the change.

Reply:
Besides, if you want to change the algorithm, that's ok. But this PR is not about changing the algorithm, it's about making isCalleeDestroyingArgs work.

Reply:
> As to pseudocode, it is there in the comments preceding the change.

So that comment needs updating to reflect this.

```
Generates:
  __swap = this; this = s; __swap.dtor();  // callee destructs `this`
or:
  __swap = this; this = s; s = __swap;  // caller destructs `this`
```

The newly added comment

> caller destroys s, so copy contents of swap back into s

Is subtly misleading, I took "back into" to mean that `__swap` was initialiazed with `s` in the first place.